### PR TITLE
fix(subscriptions): allow zero dollar invoices

### PR DESCRIPTION
**Pull request by StaberindeZA**
## Because

- For invoices where  the amount due is less than Stripe minimums, the
  subscription is successfully created, however the user is shown an
  error message on completion. This applies to both Stripe and PayPal.

## This pull request

- Updates the PayPal zero amount invoice logic.
- Updates the Stripe logic not to expect a payment intent.

## Issue that this pull request solves

Closes: #11573 

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

Paypal 100% off ($0.00 invoice)
image

Paypal discount puts amount beneath minimum amount ($0.50)
image

Credit Card 100% off ($0.00 invoice)
image

Credit Card discount puts amount beneath minimum amount ($0.50)
image

## Other information (Optional)

Any other information that is important to this pull request.

### CODE REVIEW

**Comment by StaberindeZA** [comment]
After spending some time investigating this, both Credit Card and Paypal payments had different reasons why $0.00 invoices were failing.

- Paypal payments
  - The processZeroInvoice logic was only called when `amount_due` was 0. However, zero invoices also occur if the `amount_due` is less than the Stripe minimum allowed amounts. (e.g $0.50)
  - processZeroInvoice logic called both `finalizeInvoice` and `payInvoiceOutOfBand`. This ended up throwing an error on attempting to pay the invoice, since it's already in status paid. Seems when an invoice doesn't require a payment, it automatically marks the status as paid. [Mentioned here](https://stripe.com/docs/billing/invoices/subscription#sub-invoice-lifecycle).

- Credit Card payments
  - The UI assumes that on a successful payment, the `latest_invoice` will have a `payment_intent`. However in the case of a zero invoice, there won't be a payment intent, which is why the UI was throwing an error.

File: packages/fxa-auth-server/lib/payments/paypal-processor.ts [hunk 1]
```diff
@@ -222,7 +223,8 @@ export class PaypalProcessor {
    */
   private async makePaymentAttempt(invoice: Stripe.Invoice) {
     const customer = invoice.customer as Stripe.Customer;
-    if (invoice.amount_due === 0) {
+    const minAmount = STRIPE_MINIMUM_CHARGE_AMOUNTS[invoice.currency] || 50;
```
Comment (chenba) [inline packages/fxa-auth-server/lib/payments/paypal-processor.ts:226]: This `50` seems arbitrary.  If we cannot find a min amount, then process the invoice.  If we get an error, then we'll know what needs to be fixed.

File: packages/fxa-payments-server/src/components/CouponForm/index.test.tsx [hunk 2]
```diff
@@ -1,11 +1,33 @@
...
 import '@testing-library/jest-dom/extend-expect';
-import CouponForm from './index';
-import { Coupon } from '../../lib/Coupon';
-import { SELECTED_PLAN } from '../../lib/mock-data';
+
+import { cleanup, fireEvent, render } from '@testing-library/react';
+import React from 'react';
 import waitForExpect from 'wait-for-expect';
 
+import CouponForm from '.';
```
Comment (chenba) [inline packages/fxa-payments-server/src/components/CouponForm/index.test.tsx:7]: TIL

File: packages/fxa-payments-server/src/components/CouponForm/index.test.tsx [hunk 3]
```diff
@@ -28,6 +50,38 @@ describe('CouponForm', () => {
...
+      expect(queryByTestId('coupon-error')).toBeInTheDocument();
+      expect(mockSetCoupon).toBeCalledWith(undefined);
+    });
+
+    const couponError = getByTestId('coupon-error');
+
+    expect(couponError).toHaveTextContent(
+      'The code you entered is invalid or expired.'
+    );
+  });
```
Comment (chenba) [inline packages/fxa-payments-server/src/components/CouponForm/index.test.tsx:78]: Thanks for adding this test.  Doesn't seem to be related to zero charge invoices, but 👍 anyway.

File: packages/fxa-auth-server/lib/payments/paypal-processor.ts (hunk 1 above)
Comment (StaberindeZA) [reply packages/fxa-auth-server/lib/payments/paypal-processor.ts:226]: Oh yes that makes more sense.

File: packages/fxa-auth-server/lib/payments/paypal.ts [hunk 4]
```diff
@@ -397,10 +397,10 @@ export class PayPalHelper {
...
    */
   public processZeroInvoice(invoice: Stripe.Invoice) {
-    return Promise.all([
-      this.stripeHelper.finalizeInvoice(invoice),
-      this.stripeHelper.payInvoiceOutOfBand(invoice),
-    ]);
+    // It appears for subscriptions that do not require payment, the invoice
+    // transitions to paid automatially.
+    // https://stripe.com/docs/billing/invoices/subscription#sub-invoice-lifecycle
+    return this.stripeHelper.finalizeInvoice(invoice);
```
[the remark this answers is not in the document]
Comment (StaberindeZA) [reply packages/fxa-auth-server/lib/payments/paypal.ts:403]: Before removing `payInvoiceOutOfBand` it would actually error on that line, with a message along the lines of "This has already been paid".

That's why I thought it's safe to remove, but I'm not sure if it applies in all scenarios with zero invoices.

File: packages/fxa-payments-server/src/components/CouponForm/index.test.tsx (hunk 2 above)
Comment (StaberindeZA) [reply packages/fxa-payments-server/src/components/CouponForm/index.test.tsx:7]: Hmm, TS Hero teaches us all. :D

File: packages/fxa-payments-server/src/components/CouponForm/index.test.tsx (hunk 3 above)
Comment (StaberindeZA) [reply packages/fxa-payments-server/src/components/CouponForm/index.test.tsx:78]: It's coming up in my next PR, so I thought might as well add it in here as well. :)

**Review by StaberindeZA — COMMENTED** [review]
I'll get started on the changes, but thought I'd leave some replies already.

File: packages/fxa-payments-server/src/components/CouponForm/index.test.tsx [hunk 5]
```diff
@@ -28,6 +50,38 @@ describe('CouponForm', () => {
...
+      );
+    };
+
+    const { queryByTestId, getByTestId } = subject();
+    fireEvent.change(getByTestId('coupon-input'), { target: { value: 'a' } });
+    fireEvent.click(getByTestId('coupon-button'));
+
+    await waitForExpect(() => {
+      expect(queryByTestId('coupon-error')).toBeInTheDocument();
+      expect(mockSetCoupon).toBeCalledWith(undefined);
```
[the remark this answers is not in the document]
Comment (StaberindeZA) [reply packages/fxa-payments-server/src/components/CouponForm/index.test.tsx:70]: Oh actually in this case we do want to ensure it's not called with any arguments.

File: packages/fxa-auth-server/lib/payments/paypal.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (chenba) [reply packages/fxa-auth-server/lib/payments/paypal.ts:403]: Ah I see.  I think it would be nice to still mark it as paid out of band (not a blocker though).

File: packages/fxa-payments-server/src/components/CouponForm/index.tsx [hunk 6]
```diff
@@ -15,23 +15,30 @@ import sentry from '../../lib/sentry';
...
       priceId: planId,
       promotionCode,
     });
 
     if (!discount?.amount) {
       throw new Error('No discount for coupon');
     }
+
+    if (total < getMinimumAmount(currency) && total !== 0) {
+      throw new Error('Total below minimum amount allowed.');
```
Comment (chenba) [inline packages/fxa-payments-server/src/components/CouponForm/index.tsx:36]: I know that this error is currently not shown the end users.  And I don't think we should.  It would be a strange one to actually show users since this is likely a misconfiguration on the price.

I'll file a jira on preventing that misconfiguration.

File: packages/fxa-auth-server/lib/payments/paypal-processor.ts [hunk 7]
```diff
@@ -222,7 +223,8 @@ export class PaypalProcessor {
    */
   private async makePaymentAttempt(invoice: Stripe.Invoice) {
     const customer = invoice.customer as Stripe.Customer;
-    if (invoice.amount_due === 0) {
+
+    if (invoice.amount_due < getMinimumAmount(invoice.currency)) {
```
Comment (chenba) [inline packages/fxa-auth-server/lib/payments/paypal-processor.ts:227]: `<=` I think, if I understand the Stripe docs correctly.

File: packages/fxa-payments-server/src/components/CouponForm/index.tsx [hunk 8]
```diff
@@ -15,23 +15,30 @@ import sentry from '../../lib/sentry';
...
+    }
+
     return discount.amount;
   } catch (err) {
+    console.log(err);
```
Comment (chenba) [inline packages/fxa-payments-server/src/components/CouponForm/index.tsx:41]: Delete?

File: packages/fxa-shared/subscriptions/stripe.ts [hunk 9]
```diff
@@ -231,3 +259,12 @@ export const getSubscriptionUpdateEligibility: (
 
   return SubscriptionUpdateEligibility.INVALID;
 };
+
+export const getMinimumAmount = (currency: string): number => {
+  if (STRIPE_MINIMUM_CHARGE_AMOUNTS.hasOwnProperty(currency))
+    return STRIPE_MINIMUM_CHARGE_AMOUNTS[currency];
```
Comment (chenba) [inline packages/fxa-shared/subscriptions/stripe.ts:265]: nit: I think we usually wrap the if block in `{}` even when there's just one statement.  (Not a convention written down anywhere though. 🤷)

File: packages/fxa-shared/subscriptions/stripe.ts [hunk 10]
```diff
@@ -231,3 +259,12 @@ export const getSubscriptionUpdateEligibility: (
 
   return SubscriptionUpdateEligibility.INVALID;
 };
+
+export const getMinimumAmount = (currency: string): number => {
+  if (STRIPE_MINIMUM_CHARGE_AMOUNTS.hasOwnProperty(currency))
```
Comment (chenba) [inline packages/fxa-shared/subscriptions/stripe.ts:264]: Any reason you prefer `hasOwnProperty` over `STRIPE_MINIMUM_CHARGE_AMOUNTS[currency]`?

File: packages/fxa-auth-server/lib/payments/paypal-processor.ts (hunk 7 above)
Comment (chenba) [reply packages/fxa-auth-server/lib/payments/paypal-processor.ts:227]: haha oops, never mind.

File: packages/fxa-payments-server/src/components/CouponForm/index.tsx (hunk 6 above)
Comment (pdehaan) [reply packages/fxa-payments-server/src/components/CouponForm/index.tsx:36]: Interesting. That should be the ESLint `curly` rule, IIRC.
https://github.com/mozilla/eslint-plugin-fxa/blob/fbcfb9db9881d644d147fa9461a4273884160553/lib/configs/default.js#L21

The ESLint `plugin:fxa/default` config should set this to `'curly': [2, 'all'],` (where `2` is old-style for "error").
I don't see why this would not be enabled via the following configs:
| ESLint CONFIG | EXTENDS
|:----|:-----
| https://github.com/mozilla/fxa/blob/main/_dev/.eslintrc | `"extends": ["plugin:fxa/recommended"],`
| https://github.com/mozilla/fxa/blob/main/packages/fxa-shared/.eslintrc | `"extends": ["plugin:fxa/recommended"],`
| https://github.com/mozilla/eslint-plugin-fxa/blob/fbcfb9db9881d644d147fa9461a4273884160553/lib/configs/recommended.js | `extends: ['plugin:fxa/default', 'prettier']`
| https://github.com/mozilla/eslint-plugin-fxa/blob/fbcfb9db9881d644d147fa9461a4273884160553/lib/configs/default.js | `extends: ['eslint:recommended'],`

And yet, looking at ESLint's `--parse-config` for packages/fxa-shared/subscriptions/stripe.ts, it looks like the `curly` rule is disabled:
```sh
npx eslint --config=_dev/.eslintrc.json --print-config packages/fxa-shared/subscriptions/stripe.ts | jq '.rules.curly' | npx prettier --parser=json
# [0, "all"]
```

So I think the only other candidate is probably the rule is being disabled by **eslint-config-prettier**:
https://github.com/prettier/eslint-config-prettier/blob/7ee41f151c7b17d2a6d5cc1211b3910712eadc1b/index.js#L10

```js
    "curly": 0,
```

File: packages/fxa-shared/subscriptions/stripe.ts (hunk 9 above)
Comment (pdehaan) [reply packages/fxa-shared/subscriptions/stripe.ts:265]: Interesting. That should be the ESLint `curly` rule, IIRC.
https://github.com/mozilla/eslint-plugin-fxa/blob/fbcfb9db9881d644d147fa9461a4273884160553/lib/configs/default.js#L21

The ESLint `plugin:fxa/default` config should set this to `'curly': [2, 'all'],` (where `2` is old-style for "error").
I don't see why this would not be enabled via the following configs:
| ESLint CONFIG | EXTENDS
|:----|:-----
| https://github.com/mozilla/fxa/blob/main/_dev/.eslintrc | `"extends": ["plugin:fxa/recommended"],`
| https://github.com/mozilla/fxa/blob/main/packages/fxa-shared/.eslintrc | `"extends": ["plugin:fxa/recommended"],`
| https://github.com/mozilla/eslint-plugin-fxa/blob/fbcfb9db9881d644d147fa9461a4273884160553/lib/configs/recommended.js | `extends: ['plugin:fxa/default', 'prettier']`
| https://github.com/mozilla/eslint-plugin-fxa/blob/fbcfb9db9881d644d147fa9461a4273884160553/lib/configs/default.js | `extends: ['eslint:recommended'],`

And yet, looking at ESLint's `--parse-config` for packages/fxa-shared/subscriptions/stripe.ts, it looks like the `curly` rule is disabled:
```sh
npx eslint --config=_dev/.eslintrc.json --print-config packages/fxa-shared/subscriptions/stripe.ts | jq '.rules.curly' | npx prettier --parser=json
# [0, "all"]
```

So I think the only other candidate is probably the rule is being disabled by **eslint-config-prettier**:
https://github.com/prettier/eslint-config-prettier/blob/7ee41f151c7b17d2a6d5cc1211b3910712eadc1b/index.js#L10

```js
    "curly": 0,
```

File: packages/fxa-shared/subscriptions/stripe.ts (hunk 10 above)
Comment (pdehaan) [reply packages/fxa-shared/subscriptions/stripe.ts:264]: I think ESLint has feelings on this as well; https://eslint.org/docs/rules/no-prototype-builtins
But interestingly, <kbd>eslint --print-config</kbd> reports that the "no-prototype-builtins" rule is set to `"error"` currently.

```sh
npx eslint --config=_dev/.eslintrc.json --print-config packages/fxa-shared/subscriptions/stripe.ts | jq --sort-keys '.rules["no-prototype-builtins"]' | npx prettier --parser=json
# ["error"]
```

**NOTE:** But I have a feeling that we're not running ESLint against .ts/.tsx files, so not sure this would have been caught by husky/lint-staged.
I need to write up a Google doc on my limited understanding of the current state of linting in this repo.

File: packages/fxa-payments-server/src/components/CouponForm/index.tsx (hunk 8 above)
Comment (StaberindeZA) [reply packages/fxa-payments-server/src/components/CouponForm/index.tsx:41]: One day I'll get a PR up without any console.logs...not today though.

File: packages/fxa-shared/subscriptions/stripe.ts (hunk 10 above)
Comment (StaberindeZA) [reply packages/fxa-shared/subscriptions/stripe.ts:264]: I was playing around trying to get JS to throw an error when it can't find a key in an object, and I think `hasOwnProperty` just stuck around from after that. (I was trying to avoid the `if` all together. Any ideas how to do that?)

I'll change it to your suggestion though. Does make it easier to read.

File: packages/fxa-auth-server/lib/payments/paypal.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (StaberindeZA) [reply packages/fxa-auth-server/lib/payments/paypal.ts:403]: Unfortunately even with `auto_advance` set to false on the finalize call, it looks like the invoice is still returned as `paid` after finalizes completes.

File: packages/fxa-shared/subscriptions/stripe.ts (hunk 9 above)
Comment (StaberindeZA) [reply packages/fxa-shared/subscriptions/stripe.ts:265]: I'll play around a bit with my `formatOnSave` settings in VSCode. That might be the cause of this.

For now I've manually fixed it.

File: packages/fxa-payments-server/src/components/CouponForm/index.tsx (hunk 6 above)
Comment (chenba) [reply packages/fxa-payments-server/src/components/CouponForm/index.tsx:36]: hmm does this need to be on the front end at all?  I believe the invoice preview endpoint can handle it.

File: packages/fxa-shared/subscriptions/stripe.ts [hunk 11]
```diff
@@ -231,3 +259,13 @@ export const getSubscriptionUpdateEligibility: (
...
 };
+
+export const getMinimumAmount = (currency: string): number => {
+  if (STRIPE_MINIMUM_CHARGE_AMOUNTS[currency]) {
+    return STRIPE_MINIMUM_CHARGE_AMOUNTS[currency];
+  }
+
+  throw new Error(
+    `Currency "${currency}" does not have a minimum amount available.`
+  );
```
Comment (chenba) [inline packages/fxa-shared/subscriptions/stripe.ts:270]: This is ok for now since I don't think anything is catching this error.  But generally speaking, determining the type of error by checking the error message text is not ideal.

**Review by chenba — CHANGES_REQUESTED** [review]
Make sense overall.  I think we can keep that logic backend only by adding a status on the response of the invoice preview endpoint.

File: packages/fxa-payments-server/src/components/CouponForm/index.tsx (hunk 6 above)
Comment (chenba) [reply packages/fxa-payments-server/src/components/CouponForm/index.tsx:36]: I wrote a semi-long winded response and then deleted it because I think I should move back a few steps: what's this validation for?  If it's just to show some message on the front end but still allow the subscription (is it still an error then?), then sure let's keep it on the front end.

Comment (StaberindeZA) [reply packages/fxa-payments-server/src/components/CouponForm/index.tsx:36]: I would say it is an error, but in the context of the Coupon.

It won't block the subscription from going through, but it will block the coupon from being applied.

File: packages/fxa-auth-server/lib/payments/stripe.ts [hunk 12]
```diff
@@ -716,18 +718,28 @@ export class StripeHelper {
...
+        const minAmount = getMinimumAmount(currency);
+        if (total !== 0 && minAmount && total < minAmount) {
+          throw error.invalidPromoCode(promotionCode);
+        }
+      } catch (error) {
+        if (
+          error instanceof AppError &&
+          error.errno === AppError.ERRNO.INVALID_PROMOTION_CODE
+        ) {
+          throw error;
```
Comment (chenba) [inline packages/fxa-auth-server/lib/payments/stripe.ts:741]: Since we are only re-throwing the invalid promotion code error here does that mean if `previewInvoice` or `getMinimumAmount` throws an error, we simply won't know about it?

File: packages/fxa-payments-server/src/lib/apiClient.ts [hunk 13]
```diff
@@ -412,10 +414,6 @@ export async function apiRetrieveCouponDetails(params: {
       { body: JSON.stringify(params) }
     );
 
-    if (!result.discountAmount) {
-      throw new Error(CouponErrorMessageType.Generic);
-    }
-
     Amplitude.coupon_FULFILLED(metricsOptions);
```
Comment (chenba) [inline packages/fxa-payments-server/src/lib/apiClient.ts:417]: Now that we don't throw in this function anymore, maybe we should clarify whether this metric mean successful request or valid coupon?

File: packages/fxa-shared/subscriptions/types.ts [hunk 14]
```diff
@@ -156,3 +156,7 @@ export const SubscriptionUpdateEligibility = {
 
 export type SubscriptionUpdateEligibility =
   typeof SubscriptionUpdateEligibility[keyof typeof SubscriptionUpdateEligibility];
+
+export enum SubscriptionStripeErrorType {
+  NO_MIN_AMOUNT_FOR_CURRENCY = 'Currency does not have a minimum amount available.',
```
Comment (chenba) [inline packages/fxa-shared/subscriptions/types.ts:161]: minimum _charge_ amount maybe?

File: packages/fxa-shared/subscriptions/types.ts [hunk 15]
```diff
@@ -156,3 +156,7 @@ export const SubscriptionUpdateEligibility = {
...
+}
```
Comment (chenba) [inline packages/fxa-shared/subscriptions/types.ts:162]: Have you considered extending `Error` to give these error a different type?

**Review by chenba — COMMENTED** [review]
Looks good overall.  There are a couple concerns on error throwing/handling changes that I'd see addressed.

File: packages/fxa-payments-server/src/lib/apiClient.ts (hunk 13 above)
Comment (StaberindeZA) [reply packages/fxa-payments-server/src/lib/apiClient.ts]: The VPN Coupons MVP doc says the following.

```
coupon_error: coupon validation failed. Capture failure type with coupon code used
coupon_success: fire when user presses apply and coupon is valid

```

It sounds like only fire the success event when the coupon is actually valid.

As part of these changes, I moved all the coupon metric events to the CouponForm component, since that's where the logic is at the moment that determines whether or not a coupon is valid.

File: packages/fxa-auth-server/lib/payments/stripe.ts (hunk 12 above)
Comment (StaberindeZA) [reply packages/fxa-auth-server/lib/payments/stripe.ts:741]: A good catch. Added an else, which will report the error to Sentry.

File: packages/fxa-auth-server/test/local/payments/stripe.js [hunk 16]
```diff
@@ -1300,6 +1349,68 @@ describe('StripeHelper', () => {
       });
       assert.deepEqual(actual, expected);
     });
+
+    it('retrieves details when previewInvoice rejects', async () => {
```
Comment (chenba) [inline packages/fxa-auth-server/test/local/payments/stripe.js:1353]: Very possible that it was just me but I found this test title confusing.  I was trying to understand _what_ is retrieving coupon details when previewInvoice rejects, when there's no retrieving call of any sort after previewInvoice in the function.

File: packages/fxa-auth-server/lib/payments/stripe.ts [hunk 17]
```diff
@@ -716,18 +718,36 @@ export class StripeHelper {
...
+        } else {
+          Sentry.withScope((scope) => {
+            scope.setContext('retrieveCouponDetails', {
+              priceId,
+              promotionCode,
+            });
+            Sentry.captureException(error);
+          });
```
Comment (chenba) [inline packages/fxa-auth-server/lib/payments/stripe.ts:749]: I cannot find a test for this new code.

File: packages/fxa-shared/subscriptions/types.ts (hunk 15 above)
Comment (chenba) [reply packages/fxa-shared/subscriptions/types.ts:169]: Not a blocker.  But I was thinking that we could use `instanceOf` to check for the type of error.  `SubscriptionStripeErrorType` seems too generic for that purpose.  E.g. `new StripeNoMinChargeCurrencyError('de')`.  Something like that.

**Review by chenba — APPROVED** [review]
Well, patches grow big so quickly these day!  😁  It looks good.  Got a few comments.  Would like to see that missing test added at least.

File: packages/fxa-auth-server/lib/payments/stripe.ts (hunk 17 above)
Comment (StaberindeZA) [reply packages/fxa-auth-server/lib/payments/stripe.ts:749]: Updated the one test to check that Sentry was called, and added another one for getMinAmount exceptions.

File: packages/fxa-auth-server/test/local/payments/stripe.js (hunk 16 above)
Comment (StaberindeZA) [reply packages/fxa-auth-server/test/local/payments/stripe.js]: I updated the text a little that should hopefully clarify it a bit. 

These are all `retrieveCouponDetails` tests. That test specifically should still return a result even though the method called by it, previewInvoice, rejects.

**Comment by StaberindeZA** [comment]
> Well, patches grow big so quickly these day! 😁 It looks good. Got a few comments. Would like to see that missing test added at least.

Thank you so much for the multiple code reviews, especially considering things changed a few times throughout this PR.
